handshake/src/types/tx.rs

```diff
@@ -0,0 +1,766 @@
+//! Handshake transaction types and associated sighash arguments.
+use std::io::{Error as IOError, Read, Write};
+use thiserror::Error;
+
+use crate::{
+    hashes::{TXID, WTXID},
+};
+
+use coins_core::{
+    hashes::{
+        blake2b256::{Blake2b256Digest, Blake2b256Writer},
+        marked::{MarkedDigest, MarkedDigestWriter}
+    },
+    ser::{ByteFormat, SerError},
+    types::tx::Transaction,
+};
+
+use crate::{
+    types::{
+        Witness,
+        LockingScript,
+        txin::{HandshakeOutpoint, HandshakeTxIn, Vin},
+        txout::{TxOut, Vout},
+        //witness::*,
+    },
+};
+
+pub trait HandshakeTransaction: Transaction {
+    /// The MarkedDigest type for the Transaction's Witness TXID
+    type WTXID: MarkedDigest<Digest = Self::Digest>;
+    /// A type that represents this transactions per-input `Witness`.
+    type Witness;
+
+    /// Instantiate a new WitnessTx from the arguments.
+    fn new<I, O, W>(version: u32, vin: I, vout: O, witnesses: W, locktime: u32) -> Self
+    where
+        I: Into<Vec<Self::TxIn>>,
+        O: Into<Vec<Self::TxOut>>,
+        W: Into<Vec<Self::Witness>>;
+
+    /// Calculates the witness txid of the transaction.
+    fn wtxid(&self) -> Self::WTXID;
+
+    /// Calculates the BIP143 sighash given the sighash args. See the
+    /// `WitnessSighashArgsSigh` documentation for more in-depth discussion of sighash.
+    fn signature_hash(
+        &self,
+        args: &Self::SighashArgs,
+    ) -> Result<Self::Digest, Self::TxError> {
+        let mut w = Self::HashWriter::default();
+        self.write_sighash_preimage(&mut w, args)?;
+        Ok(w.finish())
+    }
+}
+
+/// Wrapper enum for returning values that may be EITHER a Witness OR a Legacy tx and the type is
+/// not known in advance. While a few transaction methods have been implemented for convenience,
+/// This wrapper must be explicitly unwrapped before the tx object can be signed.
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, Eq, PartialEq)]
+pub struct HandshakeTx {
+    /// The version number. Usually 1 or 2.
+    pub(crate) version: u32,
+    /// The vector of inputs
+    pub(crate) vin: Vin,
+    /// The vector of outputs
+    pub(crate) vout: Vout,
+    /// The nLocktime field.
+    pub(crate) locktime: u32,
+    ///
+    pub(crate) witnesses: Vec<Witness>,
+
+}
+
+impl Default for HandshakeTx {
+    fn default() -> Self {
+        Self{
+            version: 0,
+            vin: vec![],
+            vout: vec![],
+            locktime: 0,
+            witnesses: vec![]
+        }
+    }
+}
+
+impl ByteFormat for HandshakeTx {
+    type Error = TxError; // Ser associated error
+
+    fn serialized_length(&self) -> usize {
+        let mut len = 4; // version
+        len += 2; // Segwit Flag
```

Comment:
This is a bug, add a test for `serialized_length`. There is no segwit flag for handshake transactions